pkg/addon/agent/controller/token.go

```diff
@@ -197,17 +225,32 @@ func buildSecret(managed *authv1alpha1.ManagedServiceAccount, caData, tokenData
 	}
 }
 
-func shouldCreateToken(managed *authv1alpha1.ManagedServiceAccount) bool {
-	if managed.Status.TokenSecretRef == nil {
-		return true
+func (r *TokenReconciler) shouldCreateToken(managed *authv1alpha1.ManagedServiceAccount, tokenSecret *corev1.Secret) (bool, error) {
+	if managed.Status.TokenSecretRef == nil || tokenSecret == nil {
+		return true, nil
 	}
+
+	// check if the token should be refreshed
 	now := metav1.Now()
 	refreshThreshold := managed.Spec.Rotation.Validity.Duration / 5 * 4
 	lifetime := managed.Status.ExpirationTimestamp.Sub(now.Time)
 	if lifetime < refreshThreshold {
-		return true
+		return true, nil
+	}
+
+	// check if the token is valid or not
+	tokenReview := &authv1.TokenReview{
```

Comment:
introduced tokenreview to check if a token is valid. We are doing it because if user deleted serviceaccount manually, and the old token is not valid any more even if we are recreating a serviceaccount.

Reply:
SGTM